universal-application-tool-0.0.1/app/services/applicant/predicate/JsonPathPredicateGenerator.java

```diff
@@ -1,20 +1,45 @@
 package services.applicant.predicate;
 
-import services.Path;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import services.applicant.exception.InvalidPredicateException;
+import services.applicant.question.ApplicantQuestion;
 import services.program.predicate.LeafOperationExpressionNode;
 
+/** Generates {@link JsonPathPredicate}s based on the current applicant filling out the program. */
 public class JsonPathPredicateGenerator {
 
+  private final ImmutableMap<Long, ApplicantQuestion> questionsById;
+
+  public JsonPathPredicateGenerator(ImmutableList<ApplicantQuestion> programQuestions) {
```

Comment:
(nit): the list _could_ have duplicates, which I think will break the collection into a map. Could this be a `set` object to guarantee uniqueness?

Reply:
(blocking) for repeated questions, each `ApplicantQuestion` will share the same question definition id. Will that cause problems?

Reply:
This should be fixed now - we use a list of `QuestionDefinition`s instead